### ISSUE TITLE
Add an overload to System.Management.Automation.Cmdlet.WriteError that accepts an System.Exception

### ISSUE THREAD

**Issue by Kriegel** [PowerShell/PowerShell#10735]
During the runtime of a advanced Function the function developer may use .NET classes which produce .NET exceptions.

To rethrow an Exception in an advanced Function we can use Write-Error, Throw or $PSCmdlet.WriteError().

To set the advanced Function as origin of the Error AND to make the .NET Exception an ErrorRecord Object $PSCmdlet.WriteError() is the best candidate.

currently the user has to make  an cumbersome switch inside the Catch block to catch .NET Exception Types and ErrorRecord Types.

```powershell
Function Get-ErrorHandling {

    [CmdletBinding()]
    Param()

    Process {
       Try {
           
           $File = Get-Item 'C:\no\such.txt' -ErrorAction 'Stop'
                      
           [double]::Parse('Hello')
             
       } Catch {

           If($_ -is [System.Management.Automation.ErrorRecord]) {
              $PSCmdlet.WriteError($_)
           } Else {
              $PSCmdlet.WriteError((New-Object System.Management.Automation.ErrorRecord -ArgumentList $_))
           }
       }
    }
}

```

So $PSCmdlet.WriteError()  needs an overload that accepts a .NET Exception and makes it into an ErrorRecord with an InvocationInfo of the advanced the Function here Named "Get-ErrorHandling" as source.

**Comment by vexx32** [comment]
As far as I know, the only time you'll get a raw exception is if there's a _parse_ error in the PowerShell script itself; i.e., PowerShell's own language parser throws the exception. In _most_ cases this won't surface, but if it does I believe ParseExceptions have an `ErrorRecord` property that you can make use of for a bit of an easier time with this kind of case. 🙂 

Do you know of any reproducible examples that produce an `Exception` object within the `catch` block?

**Comment by Kriegel** [comment]
Currently i have no Example. I will dig into.

If you execute the following Example you will see that the InvocationInfo is not set to the advanced function. The properties "MyCommand" and "InvocationName" of the InvocationInfo are Empty !

To set the InvocationInfo to $PSCmdlet.MyInvocation is one part of me suggestion.

```powershell
Function Get-ErrorHandling {

    [CmdletBinding()]
    Param()

    Process {
       Try {

              [double]::Parse('Hello')

       } Catch {

            $PSCmdlet.WriteError($_)
       }
    }
}

Get-ErrorHandling

$Error[0] | select -ExpandProperty InvocationInfo | select *

```

**Comment by vexx32** [comment]
I was actually under the impression that was already done, interesting. In most cases I've seen, I'm pretty sure the InvocationInfo / MyCommand information is set when you call `$PSCmdlet.WriteError()`
😕

**Comment by Kriegel** [comment]
There is a very lenghtly article in [PowerShell Magazin to create Cusom ErrorRecords](https://www.powershellmagazine.com/2011/09/14/custom-errors/) 
I think PowerShell schould do the heavy lifting for the user.

this will fail:
```powershell
Function Get-ErrorHandling {

    [CmdletBinding()]
    Param()

    Process {
       $PSCmdlet.WriteError(([Exception]::new("error message")))
    }
}

Get-ErrorHandling
```
even in this blog post the User must create an custom ErrorRecord before using the ThrowTerminatingError() Method ;-)
[https://vexx32.github.io/2019/01/31/PowerShell-Error-Handling/](https://vexx32.github.io/2019/01/31/PowerShell-Error-Handling/ )

**Comment by SeeminglyScience** [comment]
@Kriegel in that example, what would the error ID, category and target object be?

**Comment by Kriegel** [comment]
@ SeeminglyScience  

You can answer your question by yourself, If you run the following Code you will see what is inside
```powershell 
Write-Error -Exception (([Exception]::new("error message"))) 2>&1 | Select-Object *
```

$PSCmdlet.WriteError() should have overloads to fill out even this Properties.
Like i said a very similar Interface like the Write-Error Cmdlet archived with overloads.

Fully overloads in pseudo-code may look like so:
```powershell
# ErrorRecord 
 $PsCmdlet.WriteError(System.Management.Automation.ErrorRecord  ErrorRecord, string  CategoryReason, string  CategoryTargetName, string  CategoryActivity, string  CategoryTargetType, string  RecommendedAction)
 	    
 # with Exception
 $PsCmdlet.WriteError(System.Exception  Exception, string  Message, System.Management.Automation.ErrorCategory  Category, string  ErrorId, System.Object  TargetObject, string  RecommendedAction, string  CategoryActivity, string  CategoryReason, string  CategoryTargetName, string  CategoryTargetType)
 
 # no Exception
 $PsCmdlet.WriteError(string  Message, System.Management.Automation.ErrorCategory  Category, string  ErrorId, System.Object  TargetObject, string  RecommendedAction, string  CategoryActivity, string  CategoryReason, string  CategoryTargetName, string  CategoryTargetType)
``` 

Ideally the User can even control which InvocationInfo he will provide with all Write* cmdlets and Functions.
But that is out of scope here.

**Comment by SeeminglyScience** [comment]
Yeah I know, the point I was trying (poorly) to make was that those details are lost (or rather replaced with meaningless defaults).  If such an overload was added, it could lead to a significant amount of first time binary module writers skipping all of those details completely.

In general my opinion is that if you want to emit a "proper" error (one that reflects the users context) then those details should be specified.  I would like to see an overload like `WriteError(Exception,string,ErrorCategory,object)` that just creates the `ErrorRecord` for you while still requiring the information.

Also I don't really see a problem with adding a switch to `Write-Error` to change context.

**Comment by Kriegel** [comment]
@SeeminglyScience  Whats driving me since PS 2.0 is Logging.

From a Logging perspective we only need a message.  (lol)

Ok... + Timestamp and Source (like script name) perhaps + severity (which here can be Error or Critical for fatal errors or Warning for non terminating Errors).

Which Propertie Data an ErrorRecord consumer needs can vary much ...
A developer may need the inner exception (which is not allways there)
For Credential Errors, a security guy may need the Username and Windows Process / Thread
A Monitoring system may need  ..... 

That is a reason to create customized  ErrorRecords.

Currently the Automatic created ErrorRecords mostly fullfill developer debuging needs.
Which is good.

- ErrorID
it is only a string.
Like a Tag to search for or to trigger something?
Therfor you can also use the Exception type and / or (Key-)Words inside the self made message.
(Message is also simply a string ;-) ) 

- Category
is an enum with 32 members does they fit all needs or do they are meanigfull for everything?
Are they also simply Tags ?

- TargetObject
is not allways applicable.
You can even create shouldprocess queries without target object ... 

do not make the corset to tight.

in gennrell I am with you! Correct filled properties are better.

With `Write-Error` the open land is allready here!

**Comment by mklement0** [comment]
I think we need to sort out the various issues here:

* To my knowledge, `$_` in `catch` block is _always_ a `System.Management.Automation.ErrorRecord` instance. 
   * That is what you, @vexx32, mostly stated. I don't think parse errors come into play here, because any code that can't be parsed doesn't get to even _execute_, and code that traps a parse error from the outside again sees a `System.Management.Automation.ErrorRecord` instance.

* **Writing this `ErrorRecord` instance to the error stream with `$PSCmdlet.WriteError($_)` or `Write-Error -ErrorRecord $_` essentially _passes it through_ - with whatever invocation info it contains - and _PowerShell should already have supplied the appropriate invocation information_.**

   * If you didn't catch the error and let an exception become a statement-terminating error, you would see the very same `ErrorRecord` instance in `$Error[0]` afterwards.

* It may be surprising that an exception that occurs during a .NET method invocation is (implicitly) wrapped in an `ErrorRecord` instance that _doesn't_ fill in the the invocation info's `.InvocationName` or `.MyCommand` properties, but that is really _a separate issue_ - note that **the _script-level_ invocation information is always there, however**.

I can't personally speak to what aspects of the current behavior are by design or not, but here's my understanding of it:

* Cmdlet-originated errors - whether nonterminating or terminating - result in `.MyCommand` being filled in - as a `System.Management.Automation.CmdletInfo` instance - but not `.InvocationName`

* All other _terminating errors_ - runtime exceptions such as `1/0`, method exceptions such as `[int]::Parse('no')`, and `throw` statements, result in neither `.MyCommand` nor `.InvocationName` being filled in.

*  `Write-Error` errors from within a function _do_ reflect the function name in `.MyCommand` - as a `System.Management.Automation.FunctionInfo`  instance  - and whatever alias name may have been used during invocation in `.InvocationName` (as a string).

However, **in _all_ cases the function context _is_ available (albeit somewhat informally), namely via  the `.ScriptStackTrace` property**.

---

In short:

* There is generally no need to construct your own `ErrorRecord` instance in a `catch` block

* If the invocation info that _PowerShell already implicitly adds_ when it wraps exceptions in `ErrorRecord`'s is lacking, we should discuss that _separately_.

---

P.S.: Curiously, `Write-Error -Exception $_.Exception` seems to behave no differently than `Write-Error -ErrorRecord $_`, though `Write-Error -Exception $_.Exception.InnerException` does seem to create a new `ErrorRecord` that suggest the enclosing function as the source - not sure what's going on there.

**Comment by Kriegel** [comment]
see also #3190 Implement New-ErrorRecord Cmdlet

**Comment by mklement0** [comment]
Sorry, I had missed part of this discussion before I posted my previous comment.

Note that advanced functions shouldn't really have to call `$PSCmdlet.WriteError()` directly - the only good reason to do that it currently is to work around the `Write-Error` bug where `$?` isn't set to `$false` in the caller's context - see #3629

That leaves us with `Write-Error`, which already _does_ permit construction of custom `ErrorRecord` instances (as stated, `Write-Error -ErrorRecord $_` simply passes the error record through).

While `Write-Error` already allows you to control _most_ aspects of the new custom `ErrorRecord` - `-ErrorId`, `-Category`, ..., the invocation info is _implicitly_ the calling site - which is sensible.

I can't speak to the need to also control the invocation info explicitly, but if there is a need, it would stand to reason to introduce a new `Write-Error` parameter to do so, as @SeeminglyScience has already suggested.

As for #3190: Similarly, adding a `-AsTerminatingError` switch, as (in essence) suggested there by @JayKul  - or `-Throw`, as proposed there by you just now - to `Write-Error`, sounds like the right solution.

---

However, there are some oddities, as previously mentioned:

* `.MyCommand` should arguably reflect the enclosing function if a terminating error occurred directly inside a function - again, this relates to the _implicit_ creation of `ErrorRecord` instances.

* I've alluded it before, but there seems to be a bug if you use `Write-Error -Exception $_.Exception ...`, which unexpectedly behaves like `Write-Error -ErrorRecord $_` and ignores other parameters such as `-Category`.

  * In short: wrapping the original exception in a custom `ErrorRecord` is currently broken, though, curiously it seems to work with `$_.Exception.InnerException`.

**Comment by Kriegel** [comment]
@mklement0 
Does that mean a .NET developer should also use a call to the Microsoft.PowerShell.Commands.WriteErrorCommand to throw an ErrorRecord?

And for what do we need the cmdlet.WriteError() Method then?

I am the biggest fan to remove the multi-variant, multi-route.
I PowerShell User can ["Overwrite" (Proxy) the Write-xxx Cmdlets](https://www.powershellgallery.com/packages/PSLog/5.0.0.200) but not the  WriteXxxx() Methods. 

But this will be a breaking change.

I think we need an consistent interface and behavior for all. PowerShell Users and .NET developers. 

So why not call the same base code with the WriteXxxx() Method and the Microsoft.PowerShell.Commands.WriteXxxxCommand ? AND Provide applicable overloads with the WriteXxxx() Methods.
I think this can be done without breaking something?


@SeeminglyScience 

As an German guy, I see many, many scripts to fail in an international environment.
Even see bad Example Scripts in the Net that search for the "Admistrators" Group by Name. 
Or PerformanceCounter naming Horror.....
I have written an RoboCopy log-file parser .... it is Multilingual   .....
(I jump crazy laughing  into the running chainsaw ...)

We need an localization system that not disservice following workflows.

I think Microsoft best bed is to NOT translate.
(this saves money for them... like to stop to translate the PowerShell help-files ....)

> Message is meant to be localized so you can't depend on it's contents.
Who translate my custom Message? It is translated automatically?

**Comment by SeeminglyScience** [comment]
> We need an localization system that not disservice following workflows.

There's a lot of folks who aren't writing scripts with localization in mind, but that's sort of a culture problem in my opinion.  Lots of sysadmins don't need to care about localization, they're writing scripts for their organization which in a lot of cases will only be local.  It's hard to all of a sudden consider all of these things immediately upon publishing your first tool.  That and most PowerShell users tend to not report issues, so if your tool does get a little bit of reach, it's nearly impossible to tell how well it works with other language settings/keyboard layouts/etc.

The localization system is mostly fine, just some folks don't use it.  Making it harder to use it isn't going to help that situation though.

> For this ErrorID and Category can help MUCH.
But ErrorID is a free text and both can also be empty.

Even so, currently you have to explicitly decide to give useless values.  If `WriteError` took an `Exception` it would look an acceptable practice.

> I think Microsoft best bed is to NOT translate.
(this saves money for them... like to stop to translate the PowerShell help-files ....)

I don't agree with that, but really it's not something up for debate.  They're going to, and they're not going to rip out the whole localization system 🤷‍♂ 

>Who translate my custom Message? It is translated automatically?

No it's not.  The point is that a consumer of an error record cannot depend on the content of `Message`.  If you are consuming your *own* error messages for logging purposes, I'd highly recommend looking into a library that is actually meant for logging like [Serilog](https://github.com/serilog/serilog).

**Comment by mklement0** [comment]
@Kriegel 

> I think we need a consistent interface and behavior for all. PowerShell Users and .NET developers.

No: Different audiences, different skill levels, different syntax.

There's plenty of duplication between PowerShell cmdlets and SDK methods already, for good reasons. `$PSCmdlet` is an awkward blurring of the lines to begin with; note that [about_Automatic_Variables]() mentions exactly _2_ of its members - `.ParameterSetName` and `.ShouldProcess()` - and you normally shouldn't need or even need to know about any others.

Yes, it make sense to _also_ provide a new `Cmdlet.WriteError()` overload _for C# developers_ who write cmdlets, to also make it easier for them to create custom `ErrorRecord` instances - see https://github.com/PowerShell/PowerShell/issues/3190#issuecomment-541078226

**Comment by Kriegel** [comment]
@mklement0 thank you for your comments here and on 3190 I learned much!
The more i dive into the Writexxxx streams the more I get the impression that the whole stream system needs an big revision.
